### ISSUE TITLE
Question - Prototyping of Cluster Series

### ISSUE THREAD

**Issue by BrianMiner** [asardaes/dtwclust#1]
I have a quick question (hope this is the best place). If we use DTW distance and then a clustering algorithm (pam or hierarchical) how to best present a prototypical series from each cluster? I have many 1000's of series so presenting all for a cluster is not feasible. Is it meaningful to look at the centroid of each cluster (mean) (average value of the series within each cluster for each time point)?

**Comment by fpetitjean** [comment]
Hi Brian, 

Yes it is - we have shown that it works extremely well using DBA as the averaging method, both for Kmeans and AHC. 

See our ICDM 2014 paper here: http://www.francois-petitjean.com/Research/Petitjean2014-ICDM-DTW.pdf

With blog post here:
http://www.kdnuggets.com/2014/12/averaging-improves-accuracy-speed-time-series-classification.html

Cheers,

- François

**Comment by BrianMiner** [comment]
Thank you! So the idea is to create the clusters and then summarize each 
cluster seperately using DBA?

**Comment by fpetitjean** [comment]
Well, the way I've done it with AHC was using the distance between centroids of clusters, so I was using the prototypes (computed by DBA) as part of the clustering process. Same for kmeans. 

Best,

- François

**Comment by BrianMiner** [comment]
I was looking for a way to represent the clusters afterwards - to check 
their viability as well as to communicate the cluster characteristics. I 
have thousands and thousands of series (customer transactions) so trying 
to boil it down to a handful of clusters with a prototype for each 
cluster was what I was thinking of. I will look at DBA for this - maybe 
what you are saying produces the same?

**Comment by BrianMiner** [comment]
I read the blog post Francois that you sent. Cool stuff! Is the idea for 
classification that we would do the following:

1) Run some sort of clustering (e.g. kmeans) on the series for each 
class separately. *Are you running the clustering on the raw data or 
after using DTW?*

2) Use DBA on each class cluster separately to get an average series 
representing that cluster (there will then be as many prototype series 
as there are clusters + classes)

3) Use DTW similarity for each query series against the prototype series 
and assign a class label to this query series depending on the class of 
the prototype it is closest to


Do I understand these steps?

thanks!!

**Comment by asardaes** [comment]
Hi Brian,

I see you already got some helpful answers so I'll just comment on your last questions. 

I'm not sure what your data looks like, but based on what you wrote, it seems that your series already have classes? Depending on what those classes mean, I suppose you could create different subsets based on those classes and then run DBA on each subset to obtain a prototype. If you believe that running a clustering algorithm on each subset gives you finer details, you can of course do so. You would then do a query as you described in point 3.

Using DBA to create the prototypes that you will then use for the NN-classifier is probably a good approach if you use DTW as your similarity measure. If you want to assess some of the shape characteristics of your data, maybe the `shape_extraction` algorithm could aid with visualization. Check the examples of that function to see what I mean.

Best,
Alexis.

**Comment by BrianMiner** [comment]
These comments are all very helpful! I apologize, in my excitement I am mixing two different tasks, both of which I am interested:

1) Clustering of (unlabeled) time series and then representing each clusters with a prototype (essentially to understand the process captured by the clustering)
2) Taking labeled series and building a model to classify future series

It sounds like the general approach to #1 is in your edit ("If you use a hierarchical algorithm, then yes, you would have to run DBA on each resulting cluster to obtain your prototypes...." or shape extraction). This would allow me extraction of prototypes for each cluster (versus a simple mean of the cluster series which doesn't work well when DTW was used to create the clusters).

For #2 the approach is to create clusters using (for example) a hierarchical clustering (DTW  distance) for each class and then summarizing each cluster using DBA. These average cluster series are used for 1-KNN

If I have this right - then I am very happy for the help!

I am curious to experiment with these versus using a recurrent neural network for classification (assuming enough and long enough series).

**Comment by asardaes** [comment]
In the first case I think you have the right idea. Be sure to look at the examples and the `predict` generic, since it could be helpful.

For your second case, if you already have labels, maybe you don't need any clustering step. That depends on your specific data and what the labels represent, but for example: if you have 2 labels, A and B, where all A series are sine waves and all B series are straight lines, you could probably directly take all A series and build a prototype with the algorithm of your choice (mean, DBA, shape extraction, etc.), and then do the same for label B.

**Comment by fpetitjean** [comment]
Hi,

Just to complete: it seems a bit from what you are saying that you're seeing clustering and use of DBA as 2 separate tasks. What distance are you using between clusters for AHC?

In any case, make sure that when you're using DTW as the space embedding your objects, then any time you use an "averaging operation", DBA should be used. 

Btw, I have used DBA for millions of time series at a time, so it shouldn't be a problem. 

Cheers,

- François

**Comment by BrianMiner** [comment]
Hello-

Yes I guess I am seeing them as separate (naively I imagine). Are you 
suggesting something like this to get the DBA created centroids 
directly/simultaneously?

train_clust <- dtwclust(train, k = 7, type = "hierarchical",method = 
"ward.D",  distance = "dtw", centroid = "dba")


When you say you processed millions, you are not talking about the 
clustering correct (the code above)? It seems that is an incredibly slow 
process from what I have experienced so far....I printed off the paper 
recommended and plan to start reading it tonight (I am in the US).

Thanks again!

**Comment by fpetitjean** [comment]
Hi Brian,

Alexis has coded the whole library, I just provided some support for DBA.
The fact that it is so greatly generic means that it might be slower but 
I'm not even sure about that.

I used it with Kmeans, coded in Java. The code is on github for DBA 
(https://github.com/fpetitjean/DBA) and on dropbox for the code of the 
paper (see the link therein). There are many tricks to make it very 
fast, mostly about memory allocation, but the main thing is that I 
wouldn't try hierarchical with more than a 1,000 series, because you 
need to have all the pairwise distances.

Best,

François

**Comment by asardaes** [comment]
I would actually like to know what were the running times when using Java and millions of series, just for comparison. Both the dtw package (not done by me) and the DBA function implemented in this package have helper functions in C, so they should be fairly quick, but not as fast as code specifically tailored to your application. 

Based on some articles I've read, depending on the application, there are cases where performing clustering in the order of minutes is considered fast, especially for massive datasets. Nevertheless, in my opinion, sometimes it's hard to get an idea of execution times based on the articles because they use different languages and machines. For example, if you were to look at the code used in the k-Shape article (you can ask the authors for it), you'd see that the dtw function they used was coded entirely in matlab, which is incredibly slow in comparison to C, so naturally it was slower by several orders of magnitude in their comparison. 

And Brian, just a heads up. Ideally, you could do some hierarchical clustering as you suggested and have it finish fast enough, but right now the `centroid` argument is ignored, so you'd have to calculate the prototypes yourself. I've added the option to provide a function to do the prototype extraction (check the latests commits), but you'll need to install the package from github to get that functionality for now.

**Comment by fpetitjean** [comment]
I would say that a Kmeans with 20 clusters and 1million time series 
would finish within a day. Does that help?

**Comment by asardaes** [comment]
Thanks, I think it gives an idea of what to expect.
